File: SMILE/tutorials/python/tutorial8.py
import pysmile

# Tutorial8 loads continuous model from the XDSL file written by Tutorial7,
# then adds discrete nodes to create a hybrid model. Inference is performed
# and model is saved to disk.

class Tutorial8:
    def __init__(self):
        print("Starting Tutorial8...")
        net = pysmile.Network()

        net.read_file("tutorial7.xdsl")

        self.create_cpt_node(net,
                        "zone", "Climate Zone",
                        ["Temperate", "Desert"],
                        60, 20)
        toa_handle = net.get_node("toa")
        net.set_node_equation("toa", 
                        'toa=If(zone="Desert",Normal(22,5),Normal(11,10))')

        perceived_handle = self.create_cpt_node(net,
                                        "perceived", "Perceived Temperature",
                                        ["Hot", "Warm", "Cold"],
                                        60, 300)
        
        net.add_arc(toa_handle, perceived_handle)

        perceived_probs = [
            0,    # P(perceived=Hot |toa in -10..0)
            0.02, # P(perceived=Warm|toa in -10..0)
            0.98, # P(perceived=Cold|toa in -10..0)
            0.05, # P(perceived=Hot |toa in 0..10)
            0.15, # P(perceived=Warm|toa in 0..10)
            0.80, # P(perceived=Cold|toa in 0..10)
            0.10, # P(perceived=Hot |toa in 10..20)
            0.80, # P(perceived=Warm|toa in 10..20)
            0.10, # P(perceived=Cold|toa in 10..20)
            0.80, # P(perceived=Hot |toa in 20..30)
            0.15, # P(perceived=Warm|toa in 20..30)
            0.05, # P(perceived=Cold|toa in 20..30)
            0.98, # P(perceived=Hot |toa in 30..40)
            0.02, # P(perceived=Warm|toa in 30..40)
            0     # P(perceived=Cold|toa in 30..40)
        ]

        net.set_node_definition(perceived_handle, perceived_probs)

        net.set_evidence("zone", "Temperate")
        print("Results in temperate zone:")
        self.update_and_show_stats(net)

        net.set_evidence("zone", "Desert")
        print("Results in desert zone:")
        self.update_and_show_stats(net)

        net.write_file("tutorial8.xdsl")
        print("Tutorial8 complete: Network written to tutorial8.xdsl")


    def show_stats(self, net, node_handle):
        node_id = net.get_node_id(node_handle)

        if net.is_evidence(node_handle):
            v = net.get_cont_evidence(node_handle)
            print(node_id + " has evidence set " + str(v))
            return

        if net.is_value_discretized(node_handle):
            print(node_id + " is discretized.")
            iv = net.get_node_equation_discretization(node_handle)
            bounds = net.get_node_equation_bounds(node_handle)
            disc_beliefs = net.get_node_value(node_handle)
            lo = bounds[0]
            for i in range(0, len(disc_beliefs)):
                hi = iv[i].boundary
                print("\tP(" + node_id + " in " + str(lo) + ".." + str(hi) 
                      + ")=" + str(disc_beliefs[i]))
                lo = hi
        else:
            stats = net.get_node_sample_stats(node_handle)
            print(node_id + ": mean=" + str(stats[0]) + " stddev=" 
                  + str(stats[1]) + " min=" + str(stats[2]) + " max=" 
                  + str(stats[3]))


    def update_and_show_stats(self, net):
        net.update_beliefs()
        for h in net.get_all_nodes():
            if net.get_node_type(h) == pysmile.NodeType.EQUATION:
                self.show_stats(net, h)

        print("")


    def create_cpt_node(self, net, id, name, outcomes, x_pos, y_pos):
        handle = net.add_node(pysmile.NodeType.CPT, id)
        
        net.set_node_name(handle, name)
        net.set_node_position(handle, x_pos, y_pos, 85, 55)
        
        initial_outcome_count = net.get_outcome_count(handle)
        
        for i in range(0, initial_outcome_count):
            net.set_outcome_id(handle, i, outcomes[i])
        
        for i in range(initial_outcome_count, len(outcomes)):
            net.add_outcome(handle, outcomes[i])
            
        return handle

File: SMILE/tutorials/csharp/Tutorial8.cs
using System;
using Smile;

// Tutorial8 loads continuous model from the XDSL file written by Tutorial7,
// then adds discrete nodes to create a hybrid model. Inference is performed
// and model is saved to disk.

namespace SmileNetTutorial
{
    class Tutorial8
    {
        public static void Run()
        {
            Console.WriteLine("Starting Tutorial8...");
            Network net = new Network();

            net.ReadFile("tutorial7.xdsl");

            CreateCptNode(
                net, "zone", "Climate Zone",
                new String[] { "Temperate", "Desert" },
                60, 20);

            int toaHandle = net.GetNode("toa");
            net.SetNodeEquation(toaHandle, 
                "toa=If(zone=\"Desert\",Normal(22,5),Normal(11,10))");

            int perceivedHandle = CreateCptNode(
                net, "perceived", "Perceived Temperature",
                new String[] { "Hot", "Warm", "Cold" },
                60, 300);
            net.AddArc(toaHandle, perceivedHandle);

            double[] perceivedProbs = new double[] {
                0,    // P(perceived=Hot |toa in -10..0)
                0.02, // P(perceived=Warm|toa in -10..0)
                0.98, // P(perceived=Cold|toa in -10..0)
                0.05, // P(perceived=Hot |toa in 0..10)
                0.15, // P(perceived=Warm|toa in 0..10)
                0.80, // P(perceived=Cold|toa in 0..10)
                0.10, // P(perceived=Hot |toa in 10..20)
                0.80, // P(perceived=Warm|toa in 10..20)
                0.10, // P(perceived=Cold|toa in 10..20)
                0.80, // P(perceived=Hot |toa in 20..30)
                0.15, // P(perceived=Warm|toa in 20..30)
                0.05, // P(perceived=Cold|toa in 20..30)
                0.98, // P(perceived=Hot |toa in 30..40)
                0.02, // P(perceived=Warm|toa in 30..40)
                0     // P(perceived=Cold|toa in 30..40)
            };
            net.SetNodeDefinition(perceivedHandle, perceivedProbs);

            net.SetEvidence("zone", "Temperate");
            Console.WriteLine("Results in temperate zone:");
            UpdateAndShowStats(net);

            net.SetEvidence("zone", "Desert");
            Console.WriteLine("Results in desert zone:\n");
            UpdateAndShowStats(net);

            net.WriteFile("tutorial8.xdsl");
            Console.WriteLine(
                "Tutorial8 complete: Network written to tutorial8.xdsl");
        }


        private static int CreateEquationNode(
            Network net, String id, String name,
            String equation, double loBound, double hiBound,
            int xPos, int yPos)
        {
            int handle = net.AddNode(Network.NodeType.Equation, id);
            net.SetNodeName(handle, name);
            net.SetNodePosition(handle, xPos, yPos, 85, 55);
            net.SetNodeEquation(handle, equation);
            net.SetNodeEquationBounds(handle, loBound, hiBound);
            return handle;
        }


        private static void ShowStats(Network net, int nodeHandle)
        {
            String nodeId = net.GetNodeId(nodeHandle);

            if (net.IsEvidence(nodeHandle))
            {
                double v = net.GetContEvidence(nodeHandle);
                Console.WriteLine("{0} has evidence set ({1})", nodeId, v);
                return;
            }

            if (net.IsValueDiscretized(nodeHandle))
            {
                Console.WriteLine("{0} is discretized.", nodeId);
                DiscretizationInterval[] iv = 
                    net.GetNodeEquationDiscretization(nodeHandle);
                double[] bounds = net.GetNodeEquationBounds(nodeHandle);
                double[] discBeliefs = net.GetNodeValue(nodeHandle);
                double lo = bounds[0];
                for (int i = 0; i < discBeliefs.Length; i++)
                {
                    double hi = iv[i].boundary;
                    Console.WriteLine("\tP({0} in {1}..{2})={3}", 
                        nodeId, lo, hi, discBeliefs[i]);
                    lo = hi;
                }
            }
            else
            {
                double[] stats = net.GetNodeSampleStats(nodeHandle);
                Console.WriteLine("{0}: mean={1} stddev={2} min={3} max={4}",
                    nodeId, stats[0], stats[1], stats[2], stats[3]);
            }
        }


        private static void UpdateAndShowStats(Network net)
        {
            net.UpdateBeliefs();
            for (int h = net.GetFirstNode(); h >= 0; h = net.GetNextNode(h))
            {
                if (net.GetNodeType(h) == Network.NodeType.Equation)
                {
                    ShowStats(net, h);
                }
            }
            Console.WriteLine();
        }


        private static int CreateCptNode(
            Network net, String id, String name,
            String[] outcomes, int xPos, int yPos)
        {
            int handle = net.AddNode(Network.NodeType.Cpt, id);

            net.SetNodeName(handle, name);
            net.SetNodePosition(handle, xPos, yPos, 85, 55);

            int initialOutcomeCount = net.GetOutcomeCount(handle);
            for (int i = 0; i < initialOutcomeCount; i++)
            {
                net.SetOutcomeId(handle, i, outcomes[i]);
            }

            for (int i = initialOutcomeCount; i < outcomes.Length; i++)
            {
                net.AddOutcome(handle, outcomes[i]);
            }

            return handle;
        }

    }
}
